File: internal/cmd/reset.go
package cmd

import (
	"fmt"
	"os"

	"github.com/dkmnx/ply/internal/fs"
	"github.com/dkmnx/ply/internal/keys"
	"github.com/spf13/cobra"
)

var resetCmd = &cobra.Command{
	Use:   "reset",
	Short: "Reset ply configuration",
	Long:  `Reset removes all encrypted data and keys from ply.

This is useful when the encryption key is lost or corrupted.
You will need to re-add your providers after running this command.`,
	Run:   runReset,
}

func init() {
	rootCmd.AddCommand(resetCmd)
}

func runReset(cmd *cobra.Command, args []string) {
	fmt.Println("This will delete:")
	fmt.Println("  - All encrypted API keys")
	fmt.Println("  - Master key from OS keyring")
	fmt.Println("  - Master key file (if exists)")
	fmt.Println()
	fmt.Println("⚠️  This action cannot be undone!")
	fmt.Println()

	// Prompt for confirmation
	fmt.Print("Are you sure you want to reset? Type 'yes' to confirm: ")
	var confirm string
	fmt.Scanln(&confirm)

	if confirm != "yes" {
		fmt.Println("Reset cancelled.")
		return
	}

	// Get data directory
	dataDir, err := fs.DataDir()
	if err != nil {
		cmd.Printf("Error getting data directory: %v\n", err)
		return
	}

	// Remove database
	dbPath := dataDir + "/database.json"
	if _, err := os.Stat(dbPath); err == nil {
		if err := os.Remove(dbPath); err != nil {
			cmd.Printf("Error removing database: %v\n", err)
			return
		}
		fmt.Println("✓ Database deleted")
	}

	// Remove database backup
	dbBackupPath := dataDir + "/database.json.bak"
	if _, err := os.Stat(dbBackupPath); err == nil {
		if err := os.Remove(dbBackupPath); err != nil {
			cmd.Printf("Error removing database backup: %v\n", err)
			return
		}
		fmt.Println("✓ Database backup deleted")
	}

	// Remove master key
	keyMgr := keys.New(dataDir)
	if err := keyMgr.Delete(); err != nil {
		cmd.Printf("Error removing master key: %v\n", err)
		return
	}
	fmt.Println("✓ Master key deleted")

	// Remove legacy key file
	legacyKeyPath := dataDir + "/master.key"
	if _, err := os.Stat(legacyKeyPath); err == nil {
		if err := os.Remove(legacyKeyPath); err != nil {
			cmd.Printf("Error removing legacy key file: %v\n", err)
			return
		}
		fmt.Println("✓ Legacy key file deleted")
	}

	// Remove password file
	passwordFile := dataDir + "/password.bin"
	if _, err := os.Stat(passwordFile); err == nil {
		if err := os.Remove(passwordFile); err != nil {
			cmd.Printf("Error removing password file: %v\n", err)
			return
		}
		fmt.Println("✓ Password file deleted")
	}

	fmt.Println()
	fmt.Println("✓ Reset complete!")
	fmt.Println()
	fmt.Println("Next steps:")
	fmt.Println("  1. Run: ply init")
	fmt.Println("  2. Run: ply setup")
	fmt.Println("  3. Re-add your providers")
}

File: src/commands/reset.rs
//! Reset command implementation

use crate::error::{PyxError, Result};
use crate::keys::manager::KeyManager;
use crate::storage::paths::{
    database_path, master_key_path, models_cache_path, passphrase_path, providers_env_path,
    settings_path,
};
use std::fs;
use std::io::IsTerminal;
use std::path::{Path, PathBuf};

/// Append `.bak` to a path (backup naming convention)
fn backup_path(path: &Path) -> PathBuf {
    let mut name = path.as_os_str().to_owned();
    name.push(".bak");
    PathBuf::from(name)
}

/// Execute the reset command
pub fn execute(skip_confirm: bool) -> Result<()> {
    eprintln!("=== Pyx Reset ===");
    eprintln!();
    eprintln!("WARNING: This will permanently delete all pyx data:");
    eprintln!("  - Master key");
    eprintln!("  - All stored API keys");
    eprintln!("  - Models cache");
    eprintln!("  - Settings");
    eprintln!();

    // Confirm deletion
    if !skip_confirm && !confirm_reset()? {
        eprintln!("Reset cancelled.");
        return Err(PyxError::Cancelled);
    }

    eprintln!();
    eprintln!("Deleting pyx data...");

    // Delete master key
    delete_file("Master key", &master_key_path()?)?;

    // Delete database
    let db_path = database_path()?;
    delete_file("Provider database", &db_path)?;
    delete_file("Database backup", &backup_path(&db_path))?;

    // Delete models cache
    delete_file("Models cache", &models_cache_path()?)?;

    // Delete settings
    let settings_p = settings_path()?;
    delete_file("Settings", &settings_p)?;
    delete_file("Settings backup", &backup_path(&settings_p))?;

    // Delete providers config
    delete_file("Providers config", &providers_env_path()?)?;

    // Delete passphrase fallback file (if exists)
    delete_file("Passphrase file", &passphrase_path()?)?;

    // Clear keyring
    if KeyManager::master_key_exists() || keyring_has_entry() {
        eprintln!("Clearing keyring entry...");
        KeyManager::clear_passphrase().unwrap_or_else(|e| {
            eprintln!("Warning: Failed to clear keyring: {e}");
        });
    }

    eprintln!();
    eprintln!("✓ Reset complete!");
    eprintln!();
    eprintln!("Pyx has been reset to initial state.");
    eprintln!("Run 'pyx init' to initialize again.");

    Ok(())
}

fn confirm_reset() -> Result<bool> {
    use std::io::stdin;

    if !stdin().is_terminal() {
        return Err(PyxError::Validation(
            "Confirmation requires an interactive terminal. Use --yes to skip.".into(),
        ));
    }

    use inquire::Confirm;

    let confirmed = Confirm::new("Are you sure you want to reset?")
        .with_default(false)
        .prompt()
        .map_err(|e| PyxError::Validation(format!("Failed to read confirmation: {e}")))?;

    Ok(confirmed)
}

fn delete_file(description: &str, path: &Path) -> Result<()> {
    if path.exists() {
        fs::remove_file(path)
            .map_err(|e| PyxError::Config(format!("Failed to delete {description}: {e}")))?;
        eprintln!("  ✓ Deleted: {description}");
    } else {
        eprintln!("  - Not found: {description}");
    }
    Ok(())
}

fn keyring_has_entry() -> bool {
    crate::keys::keyring::has_entry()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_delete_file() {
        use tempfile::tempdir;

        let dir = tempdir().unwrap();
        let path = dir.path().join("test.txt");

        fs::write(&path, "test").unwrap();
        assert!(path.exists());

        delete_file("Test file", &path).unwrap();
        assert!(!path.exists());

        assert!(delete_file("Non-existent", &path).is_ok());
    }

    #[test]
    fn test_backup_path() {
        let path = PathBuf::from("/data/database.json");
        let bak = backup_path(&path);
        assert_eq!(bak, PathBuf::from("/data/database.json.bak"));
    }

    #[test]
    fn test_backup_path_no_extension() {
        let path = PathBuf::from("/data/db");
        let bak = backup_path(&path);
        assert_eq!(bak, PathBuf::from("/data/db.bak"));
    }
}
